WordPress/gutenberg
# Navigation Block: Use dom.focus for focus control

Fixes #57359

## What?

This PR avoids using `querySelector`/`querySelectorAll` and uses `focus` from the `@wordpress/dom` package for focus control in the navigation dialog. This fixes a focus error when the Navigation block has a search block, as reported in #57359.

## Why?

The following two things are expected in the Navigation block dialog:

1. Focus on the first focusable element when opening a dialog
2. Restrict focus movement to the dialog when pressing the Tab key

Regarding the first point, it tries to find focusable elements in the dialog starting from the element with the `.wp-block-navigation-item` class. However, if it is just a Search block, there is no element with this class, so focus will fail and you will get a console error.

## How?

~~The `@wordpress/dom` package has a `focus` object that finds focusable elements. I used this and replaced `querySelector`/`querySelectorAll`. This should always focus the appropriate first element, no matter what blocks the navigation block contains.~~

Update: To avoid the increase in bundle size, I fixed the logic of a problem (searching for the first focusing element in the dialog) without using the `@wordpress/dom` package.
​
I also moved the `focusFirstElement` callback to the content wrapper div (`.wp-block-navigation__responsive-container-content`) to accurately locate the first element (not including the close button) when the dialog opens.

Before:

```html
<nav class="wp-block-navigation>
  <button class="wp-block-navigation__responsive-container-open"></button>
  <div class="wp-block-navigation__responsive-container">
    <div class="wp-block-navigation__responsive-close">
      <div class="wp-block-navigation__responsive-dialog" data-wp-watch="callbacks.focusFirstElement">
        <button aria-label="Close menu" class="wp-block-navigation__responsive-container-close"><</button>
        <div class="wp-block-navigation__responsive-container-content">
          
        </div>
      </div>
    </div>
  </div>
</nav>
```

After:

```html
<nav class="wp-block-navigation>
  <button class="wp-block-navigation__responsive-container-open"></button>
  <div class="wp-block-navigation__responsive-container">
    <div class="wp-block-navigation__responsive-close">
      <div class="wp-block-navigation__responsive-dialog">
        <button class="wp-block-navigation__responsive-container-close"></button>
        <div class="wp-block-navigation__responsive-container-content" data-wp-watch="callbacks.focusFirstElement">
          
        </div>
      </div>
    </div>
  </div>
</nav>
```

### Testing Instructions for Keyboard

I'm a Windows user, and I've confirmed it works in Chrome and Firefox browsers. I would be happy if you could confirm that it works as expected on Mac OS.

- Add the following two navigations to the template. In both cases, change the Overlay Menu setting to "Always".
  - Contains only a Search block
  - Contains page list or page links
- Use the `Tab` key to focus on the navigation button and press `Enter`.
  - For navigations with a list of pages, the focus should be on the first page link.
  - If the navigation has only a search block, the focus should be on the text field _inside_ the search block.
- Press the Tab key continuously.
- Focus should not leave the dialog.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/54422211/8f019046-23bb-42a0-8c4f-510c72f59b02
